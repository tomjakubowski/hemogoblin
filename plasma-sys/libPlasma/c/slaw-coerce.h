
/* (c)  oblong industries */

#ifndef SLAW_COERCE_INTERROGATION
#define SLAW_COERCE_INTERROGATION

#include "libLoam/c/ob-types.h"
#include "libPlasma/c/slaw.h"

#ifdef __cplusplus
extern "C" {
#endif

/**                 Attempts to interpret the given slaw as an unt64 if
 *                  at all possible.  In particular, if s is any atomic
 *                  numeric type, it will be converted to an unt64.
 *                  If s is a string, the string will be
 *                  parsed as an unt64 if possible.
 *                  Is s is a list of one element or an array of one element,
 *                  then that one element shall be treated to the same
 *                  opportunities for coercion as described above.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets integer if return value is OB_OK,
 *                  no guarantee otherwise
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_unt64 (bslaw s, unt64 *result);

/**                 Attempts to interpret the given slaw as an unt32 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_unt64, which see, and the result is cast to
 *                  unt32.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_unt32 (bslaw s, unt32 *result);

/**                 Attempts to interpret the given slaw as an unt16 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_unt16, which see, and the result is cast to
 *                  unt16.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_unt16 (bslaw s, unt16 *result);

/**                 Attempts to interpret the given slaw as an unt8 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_unt8, which see, and the result is cast to
 *                  unt8.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_unt8 (bslaw s, unt8 *result);

/**                 Attempts to interpret the given slaw as an int64 if
 *                  at all possible.  In particular, if s is any atomic
 *                  numeric type, it will be converted to an int64.
 *                  If s is a string, the string will be
 *                  parsed as an int64 if possible.
 *                  Is s is a list of one element or an array of one element,
 *                  then that one element shall be treated to the same
 *                  opportunities for coercion as described above.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets integer if return value is OB_OK,
 *                  no guarantee otherwise
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_int64 (bslaw s, int64 *result);

/**                 Attempts to interpret the given slaw as an int32 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_int64, which see, and the result is cast to
 *                  int32.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_int32 (bslaw s, int32 *result);

/**                 Attempts to interpret the given slaw as an int16 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_int16, which see, and the result is cast to
 *                  int16.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_int16 (bslaw s, int16 *result);

/**                 Attempts to interpret the given slaw as an int8 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_int8, which see, and the result is cast to
 *                  int8.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_int8 (bslaw s, int8 *result);

/**                 Attempts to interpret the given slaw as a float64 if
 *                  at all possible.  In particular, if s is any atomic
 *                  numeric type, it will be converted to a float64.
 *                  If s is a string, the string will be
 *                  parsed as a float64 if possible.
 *                  Is s is a list of one element or an array of one element,
 *                  then that one element shall be treated to the same
 *                  opportunities for coercion as described above.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets float if return value is OB_OK,
 *                  no guarantee otherwise
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_float64 (bslaw s, float64 *result);

/**                 Attempts to interpret the given slaw as an float32 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_int64, which see, and the result is cast to
 *                  float32.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_float32 (bslaw s, float32 *result);

/**                 Attempts to interpret the given slaw as a length-2
 *                  vector if at all possible.  In particular, if s is a
 *                  list of the proper length, then the elements of the
 *                  list will be coerced as by slaw_to_float64.
 *                  If s is either an array or a vector of any real type,
 *                  of the proper length, it will be coerced to a vector
 *                  of float64.  And if s is a string, it will be parsed
 *                  as numbers separated by commas, if possible.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets the vector if return value is OB_OK,
 *                  otherwise unchanged
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, SLAW_WRONG_LENGTH,
 *                  or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_v2 (bslaw s, v2float64 *result);

/**                 Attempts to interpret the given slaw as a length-3
 *                  vector if at all possible.  In particular, if s is a
 *                  list of the proper length, then the elements of the
 *                  list will be coerced as by slaw_to_float64.
 *                  If s is either an array or a vector of any real type,
 *                  of the proper length, it will be coerced to a vector
 *                  of float64.  And if s is a string, it will be parsed
 *                  as numbers separated by commas, if possible.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets the vector if return value is OB_OK,
 *                  otherwise unchanged
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, SLAW_WRONG_LENGTH,
 *                  or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_v3 (bslaw s, v3float64 *result);

/**                 Attempts to interpret the given slaw as a length-4
 *                  vector if at all possible.  In particular, if s is a
 *                  list of the proper length, then the elements of the
 *                  list will be coerced as by slaw_to_float64.
 *                  If s is either an array or a vector of any real type,
 *                  of the proper length, it will be coerced to a vector
 *                  of float64.  And if s is a string, it will be parsed
 *                  as numbers separated by commas, if possible.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets the vector if return value is OB_OK,
 *                  otherwise unchanged
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, SLAW_WRONG_LENGTH,
 *                  or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_v4 (bslaw s, v4float64 *result);

/**                 Implements the same coercion behavior as slaw_to_v2(),
 *                  slaw_to_v3(), and slaw_to_v4(), but for an arbitrary
 *                  length.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets the vector elements if return value is OB_OK,
 *                  otherwise unchanged
 *
 * \param[in]       capacity is the maximum number of float64s which can
 *                  be written to \a result.
 *
 * \param[out]      len is the number of vector elements actually written
 *                  to \a result, which will be less than or equal to
 *                  \a capacity.
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, SLAW_WRONG_LENGTH,
 *                  or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_vn (bslaw s, float64 *result, int capacity,
                                    int *len);


/**       Attempts to interpret the given slaw as a v ## N ## T, by
 *         first calling slaw_to_v ## N and then casting the
 *         components of the result to T
 */
#define DECLARE_V_COERCION(N, T)                                               \
  OB_PLASMA_API ob_retort slaw_to_v##N##T (bslaw s, v##N##T *result)

#define FOR_ALL_NUMERIC_TYPES(N, M)                                            \
  M (N, int8);                                                                 \
  M (N, int16);                                                                \
  M (N, int32);                                                                \
  M (N, int64);                                                                \
  M (N, unt8);                                                                 \
  M (N, unt16);                                                                \
  M (N, unt32);                                                                \
  M (N, unt64);                                                                \
  M (N, float32);                                                              \
  M (N, float64)

FOR_ALL_NUMERIC_TYPES (2, DECLARE_V_COERCION);
FOR_ALL_NUMERIC_TYPES (3, DECLARE_V_COERCION);
FOR_ALL_NUMERIC_TYPES (4, DECLARE_V_COERCION);

#undef FOR_ALL_NUMERIC_TYPES
#undef DECLARE_V_COERCION

/**                 Attempts to interpret the given slaw as a boolean if
 *                  at all possible.  In particular, if s is any atomic
 *                  numeric type with a value of "0" or "1", it will
 *                  be converted to false or true, respectively.
 *                  If s is a string, it may be either "true" or "false",
 *                  case-insensitively.
 *                  Is s is a list of one element or an array of one element,
 *                  then that one element shall be treated to the same
 *                  opportunities for coercion as described above.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets boolean if return value is OB_OK,
 *                  no guarantee otherwise
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_boolean (bslaw s, bool *result);

#ifdef __cplusplus
}
#endif

#endif /* SLAW_COERCE_INTERROGATION */
