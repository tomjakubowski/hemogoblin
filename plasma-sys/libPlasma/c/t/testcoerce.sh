#!/bin/bash
srcdir="$(cd "$(dirname "$0")" && pwd)"

# Handle internal valgrind
IV=""
if [ "$#" == "2" ]; then
    if [ "$1" == "--internal-valgrind" ]; then
        IV=$2
    fi
fi

PATH=.:$PATH

$IV \
testcoerce > scratch/testcoerce.out
[ "$?" != "0" ] && exit 1

exec diff -ub scratch/testcoerce.out $srcdir/testcoerce.expected
