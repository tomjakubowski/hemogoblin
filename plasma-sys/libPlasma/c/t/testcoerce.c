
/* (c)  oblong industries */

#include "libPlasma/c/slaw-coerce.h"
#include <stdio.h>
#include <stdlib.h>

// This is just to make valgrind happy!
static slaw to_be_freed[500];
static int free_counter = 0;

static slaw f (slaw s)
{
  to_be_freed[free_counter++] = s;
  return s;
}

static void report_unt64 (char c, unt64 x, ob_retort err)
{
  if (err == OB_OK)
    printf ("%c) %" OB_FMT_64 "u"
            "\n",
            c, x);
  else if (err == SLAW_RANGE_ERR)
    printf ("%c) SLAW_RANGE_ERR\n", c);
  else if (err == SLAW_NOT_NUMERIC)
    printf ("%c) SLAW_NOT_NUMERIC\n", c);
  else
    printf ("%c) some other error!\n", c);
}

static void report_int64 (char c, int64 y, ob_retort err)
{
  if (err == OB_OK)
    printf ("%c) %" OB_FMT_64 "d"
            "\n",
            c, y);
  else if (err == SLAW_RANGE_ERR)
    printf ("%c) SLAW_RANGE_ERR\n", c);
  else if (err == SLAW_NOT_NUMERIC)
    printf ("%c) SLAW_NOT_NUMERIC\n", c);
  else
    printf ("%c) some other error!\n", c);
}

typedef union
{
  float64 f;
  unt64 u;
} pun64;

static void report_float64 (char c, float64 z, ob_retort err)
{
  if (err == OB_OK)
    {
      // Unfortunately, printf gives different results on different plaforms:
      // printf ("%c) %f\n", c, z);
      //
      // So to avoid that, print the bitwise representation of the
      // floating point number:
      pun64 bizarre;
      bizarre.f = z;
      printf ("%c) hex bits of float64 = %016" OB_FMT_64 "X\n", c, bizarre.u);
    }
  else if (err == SLAW_RANGE_ERR)
    printf ("%c) SLAW_RANGE_ERR\n", c);
  else if (err == SLAW_NOT_NUMERIC)
    printf ("%c) SLAW_NOT_NUMERIC\n", c);
  else
    printf ("%c) some other error!\n", c);
}

int main (int argc, char **argv)
{
  unt64 x;
  int64 y;
  float64 z;
  ob_retort err;
  int i;

  err =
    slaw_to_unt64 (f (slaw_unt64 (OB_CONST_U64 (12345678901234567890))), &x);
  report_unt64 ('a', x, err);

  err = slaw_to_unt64 (f (slaw_int64 (OB_CONST_I64 (1234567890123456789))), &x);
  report_unt64 ('b', x, err);

  err =
    slaw_to_unt64 (f (slaw_int64 (OB_CONST_I64 (-1234567890123456789))), &x);
  report_unt64 ('c', x, err);

  err = slaw_to_unt64 (f (slaw_int32 (1234567890)), &x);
  report_unt64 ('d', x, err);

  err = slaw_to_unt64 (f (slaw_int32 (-1234567890)), &x);
  report_unt64 ('e', x, err);

  err = slaw_to_unt64 (f (slaw_unt32 (1234567890)), &x);
  report_unt64 ('f', x, err);

  err = slaw_to_unt64 (f (slaw_unt16 (12345)), &x);
  report_unt64 ('g', x, err);

  err = slaw_to_unt64 (f (slaw_unt8 (123)), &x);
  report_unt64 ('h', x, err);

  err = slaw_to_unt64 (f (slaw_float64 (1e19)), &x);
  report_unt64 ('i', x, err);

  err = slaw_to_unt64 (f (slaw_float32 (1.2297829e+19f)), &x);
  report_unt64 ('j', x, err);

  err = slaw_to_unt64 (f (slaw_string ("12345678901234567890")), &x);
  report_unt64 ('k', x, err);

  err = slaw_to_unt64 (f (slaw_string ("blech")), &x);
  report_unt64 ('l', x, err);

  err = slaw_to_unt64 (f (slaw_nil ()), &x);
  report_unt64 ('m', x, err);

  err = slaw_to_unt64 (f (slaw_float64 (-1.0)), &x);
  report_unt64 ('n', x, err);

  err = slaw_to_unt64 (f (slaw_float64 (1.5)), &x);
  report_unt64 ('o', x, err);

  err = slaw_to_unt64 (f (slaw_int16_array_filled (1, 157)), &x);
  report_unt64 ('p', x, err);

  err = slaw_to_unt64 (f (slaw_int8_array_filled (1, 7)), &x);
  report_unt64 ('q', x, err);

  err = slaw_to_unt64 (f (slaw_list_inline_c ("0x0015ed", NULL)), &x);
  report_unt64 ('r', x, err);

  err = slaw_to_unt64 (f (slaw_int64_array_filled (1, OB_CONST_I64 (
                                                        -4503599627370495))),
                       &x);
  report_unt64 ('s', x, err);

  err = slaw_to_unt64 (f (slaw_unt64_array_filled (1, OB_CONST_U64 (
                                                        4503599627370495))),
                       &x);
  report_unt64 ('t', x, err);

  err =
    slaw_to_unt64 (f (slaw_float64_array_filled (1, 4503599627370495.0)), &x);
  report_unt64 ('u', x, err);

  err = slaw_to_unt64 (f (slaw_float32_array_filled (1, -8388607.0)), &x);
  report_unt64 ('v', x, err);

  err = slaw_to_unt64 (f (slaw_int16 (-157)), &x);
  report_unt64 ('w', x, err);

  err = slaw_to_unt64 (f (slaw_int8 (-7)), &x);
  report_unt64 ('x', x, err);

  err = slaw_to_unt64 (f (slaw_int32_array_filled (1, -2147483647 - 1)), &x);
  report_unt64 ('y', x, err);

  err = slaw_to_unt64 (f (slaw_unt8_array_filled (1, 255)), &x);
  report_unt64 ('z', x, err);

  err = slaw_to_unt64 (f (slaw_unt16_array_filled (1, 65535)), &x);
  report_unt64 ('0', x, err);

  err = slaw_to_unt64 (f (slaw_unt32_array_filled (1, 4294967295U)), &x);
  report_unt64 ('1', x, err);

  err = slaw_to_unt64 (f (slaw_unt32_array_filled (0, 4294967295U)), &x);
  report_unt64 ('2', x, err);

  err = slaw_to_unt64 (f (slaw_unt32_array_filled (2, 4294967295U)), &x);
  report_unt64 ('3', x, err);

  printf ("\n");

  err =
    slaw_to_int64 (f (slaw_unt64 (OB_CONST_U64 (12345678901234567890))), &y);
  report_int64 ('a', y, err);

  err = slaw_to_int64 (f (slaw_int64 (OB_CONST_I64 (1234567890123456789))), &y);
  report_int64 ('b', y, err);

  err =
    slaw_to_int64 (f (slaw_int64 (OB_CONST_I64 (-1234567890123456789))), &y);
  report_int64 ('c', y, err);

  err = slaw_to_int64 (f (slaw_int32 (1234567890)), &y);
  report_int64 ('d', y, err);

  err = slaw_to_int64 (f (slaw_int32 (-1234567890)), &y);
  report_int64 ('e', y, err);

  err = slaw_to_int64 (f (slaw_unt32 (1234567890)), &y);
  report_int64 ('f', y, err);

  err = slaw_to_int64 (f (slaw_unt16 (12345)), &y);
  report_int64 ('g', y, err);

  err = slaw_to_int64 (f (slaw_unt8 (123)), &y);
  report_int64 ('h', y, err);

  err = slaw_to_int64 (f (slaw_float64 (1e19)), &y);
  report_int64 ('i', y, err);

  err = slaw_to_int64 (f (slaw_float32 (1.2297829e+19f)), &y);
  report_int64 ('j', y, err);

  err = slaw_to_int64 (f (slaw_string ("12345678901234567890")), &y);
  report_int64 ('k', y, err);

  err = slaw_to_int64 (f (slaw_string ("blech")), &y);
  report_int64 ('l', y, err);

  err = slaw_to_int64 (f (slaw_nil ()), &y);
  report_int64 ('m', y, err);

  err = slaw_to_int64 (f (slaw_float64 (-1.0)), &y);
  report_int64 ('n', y, err);

  err = slaw_to_int64 (f (slaw_float64 (1.5)), &y);
  report_int64 ('o', y, err);

  err = slaw_to_int64 (f (slaw_int16_array_filled (1, 157)), &y);
  report_int64 ('p', y, err);

  err = slaw_to_int64 (f (slaw_int8_array_filled (1, 7)), &y);
  report_int64 ('q', y, err);

  err = slaw_to_int64 (f (slaw_list_inline_c ("0x0015ed", NULL)), &y);
  report_int64 ('r', y, err);

  err = slaw_to_int64 (f (slaw_int64_array_filled (1, OB_CONST_I64 (
                                                        -4503599627370495))),
                       &y);
  report_int64 ('s', y, err);

  err = slaw_to_int64 (f (slaw_unt64_array_filled (1, OB_CONST_U64 (
                                                        4503599627370495))),
                       &y);
  report_int64 ('t', y, err);

  err =
    slaw_to_int64 (f (slaw_float64_array_filled (1, 4503599627370495.0)), &y);
  report_int64 ('u', y, err);

  err = slaw_to_int64 (f (slaw_float32_array_filled (1, -8388607.0)), &y);
  report_int64 ('v', y, err);

  err = slaw_to_int64 (f (slaw_int16 (-157)), &y);
  report_int64 ('w', y, err);

  err = slaw_to_int64 (f (slaw_int8 (-7)), &y);
  report_int64 ('x', y, err);

  err = slaw_to_int64 (f (slaw_int32_array_filled (1, -2147483647 - 1)), &y);
  report_int64 ('y', y, err);

  err = slaw_to_int64 (f (slaw_unt8_array_filled (1, 255)), &y);
  report_int64 ('z', y, err);

  err = slaw_to_int64 (f (slaw_unt16_array_filled (1, 65535)), &y);
  report_int64 ('0', y, err);

  err = slaw_to_int64 (f (slaw_unt32_array_filled (1, 4294967295U)), &y);
  report_int64 ('1', y, err);

  err = slaw_to_int64 (f (slaw_unt32_array_filled (0, 4294967295U)), &y);
  report_int64 ('2', y, err);

  err = slaw_to_int64 (f (slaw_unt32_array_filled (2, 4294967295U)), &y);
  report_int64 ('3', y, err);

  printf ("\n");

  err =
    slaw_to_float64 (f (slaw_unt64 (OB_CONST_U64 (12345678901234567890))), &z);
  report_float64 ('a', z, err);

  err =
    slaw_to_float64 (f (slaw_int64 (OB_CONST_I64 (1234567890123456789))), &z);
  report_float64 ('b', z, err);

  err =
    slaw_to_float64 (f (slaw_int64 (OB_CONST_I64 (-1234567890123456789))), &z);
  report_float64 ('c', z, err);

  err = slaw_to_float64 (f (slaw_int32 (1234567890)), &z);
  report_float64 ('d', z, err);

  err = slaw_to_float64 (f (slaw_int32 (-1234567890)), &z);
  report_float64 ('e', z, err);

  err = slaw_to_float64 (f (slaw_unt32 (1234567890)), &z);
  report_float64 ('f', z, err);

  err = slaw_to_float64 (f (slaw_unt16 (12345)), &z);
  report_float64 ('g', z, err);

  err = slaw_to_float64 (f (slaw_unt8 (123)), &z);
  report_float64 ('h', z, err);

  err = slaw_to_float64 (f (slaw_float64 (1e19)), &z);
  report_float64 ('i', z, err);

  err = slaw_to_float64 (f (slaw_float32 (1.2297829e+19f)), &z);
  report_float64 ('j', z, err);

  err = slaw_to_float64 (f (slaw_string ("12345678901234567890")), &z);
  report_float64 ('k', z, err);

  err = slaw_to_float64 (f (slaw_string ("blech")), &z);
  report_float64 ('l', z, err);

  err = slaw_to_float64 (f (slaw_nil ()), &z);
  report_float64 ('m', z, err);

  err = slaw_to_float64 (f (slaw_float64 (-1.0)), &z);
  report_float64 ('n', z, err);

  err = slaw_to_float64 (f (slaw_float64 (1.5)), &z);
  report_float64 ('o', z, err);

  err = slaw_to_float64 (f (slaw_int16_array_filled (1, 157)), &z);
  report_float64 ('p', z, err);

  err = slaw_to_float64 (f (slaw_int8_array_filled (1, 7)), &z);
  report_float64 ('q', z, err);

  err = slaw_to_float64 (f (slaw_list_inline_c ("0x0015ed", NULL)), &z);
  report_float64 ('r', z, err);

  err = slaw_to_float64 (f (slaw_int64_array_filled (1, OB_CONST_I64 (
                                                          -4503599627370495))),
                         &z);
  report_float64 ('s', z, err);

  err = slaw_to_float64 (f (slaw_unt64_array_filled (1, OB_CONST_U64 (
                                                          4503599627370495))),
                         &z);
  report_float64 ('t', z, err);

  err =
    slaw_to_float64 (f (slaw_float64_array_filled (1, 4503599627370495.0)), &z);
  report_float64 ('u', z, err);

  err = slaw_to_float64 (f (slaw_float32_array_filled (1, -8388607.0)), &z);
  report_float64 ('v', z, err);

  err = slaw_to_float64 (f (slaw_int16 (-157)), &z);
  report_float64 ('w', z, err);

  err = slaw_to_float64 (f (slaw_int8 (-7)), &z);
  report_float64 ('x', z, err);

  err = slaw_to_float64 (f (slaw_int32_array_filled (1, -2147483647 - 1)), &z);
  report_float64 ('y', z, err);

  err = slaw_to_float64 (f (slaw_unt8_array_filled (1, 255)), &z);
  report_float64 ('z', z, err);

  err = slaw_to_float64 (f (slaw_unt16_array_filled (1, 65535)), &z);
  report_float64 ('0', z, err);

  err = slaw_to_float64 (f (slaw_unt32_array_filled (1, 4294967295U)), &z);
  report_float64 ('1', z, err);

  err = slaw_to_float64 (f (slaw_unt32_array_filled (0, 4294967295U)), &z);
  report_float64 ('2', z, err);

  err = slaw_to_float64 (f (slaw_unt32_array_filled (2, 4294967295U)), &z);
  report_float64 ('3', z, err);

  // free stuff to please valgrind
  for (i = 0; i < free_counter; i++)
    slaw_free (to_be_freed[i]);

  return EXIT_SUCCESS;
}
