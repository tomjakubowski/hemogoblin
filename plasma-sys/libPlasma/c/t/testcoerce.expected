a) 12345678901234567890
b) 1234567890123456789
c) SLAW_RANGE_ERR
d) 1234567890
e) SLAW_RANGE_ERR
f) 1234567890
g) 12345
h) 123
i) 10000000000000000000
j) 12297828649465282560
k) 12345678901234567890
l) SLAW_NOT_NUMERIC
m) SLAW_NOT_NUMERIC
n) SLAW_RANGE_ERR
o) SLAW_RANGE_ERR
p) 157
q) 7
r) 5613
s) SLAW_RANGE_ERR
t) 4503599627370495
u) 4503599627370495
v) SLAW_RANGE_ERR
w) SLAW_RANGE_ERR
x) SLAW_RANGE_ERR
y) SLAW_RANGE_ERR
z) 255
0) 65535
1) 4294967295
2) SLAW_NOT_NUMERIC
3) SLAW_NOT_NUMERIC

a) SLAW_RANGE_ERR
b) 1234567890123456789
c) -1234567890123456789
d) 1234567890
e) -1234567890
f) 1234567890
g) 12345
h) 123
i) SLAW_RANGE_ERR
j) SLAW_RANGE_ERR
k) SLAW_RANGE_ERR
l) SLAW_NOT_NUMERIC
m) SLAW_NOT_NUMERIC
n) -1
o) SLAW_RANGE_ERR
p) 157
q) 7
r) 5613
s) -4503599627370495
t) 4503599627370495
u) 4503599627370495
v) -8388607
w) -157
x) -7
y) -2147483648
z) 255
0) 65535
1) 4294967295
2) SLAW_NOT_NUMERIC
3) SLAW_NOT_NUMERIC

a) SLAW_RANGE_ERR
b) SLAW_RANGE_ERR
c) SLAW_RANGE_ERR
d) hex bits of float64 = 41D26580B4800000
e) hex bits of float64 = C1D26580B4800000
f) hex bits of float64 = 41D26580B4800000
g) hex bits of float64 = 40C81C8000000000
h) hex bits of float64 = 405EC00000000000
i) hex bits of float64 = 43E158E460913D00
j) hex bits of float64 = 43E5555540000000
k) SLAW_RANGE_ERR
l) SLAW_NOT_NUMERIC
m) SLAW_NOT_NUMERIC
n) hex bits of float64 = BFF0000000000000
o) hex bits of float64 = 3FF8000000000000
p) hex bits of float64 = 4063A00000000000
q) hex bits of float64 = 401C000000000000
r) hex bits of float64 = 40B5ED0000000000
s) hex bits of float64 = C32FFFFFFFFFFFFE
t) hex bits of float64 = 432FFFFFFFFFFFFE
u) hex bits of float64 = 432FFFFFFFFFFFFE
v) hex bits of float64 = C15FFFFFC0000000
w) hex bits of float64 = C063A00000000000
x) hex bits of float64 = C01C000000000000
y) hex bits of float64 = C1E0000000000000
z) hex bits of float64 = 406FE00000000000
0) hex bits of float64 = 40EFFFE000000000
1) hex bits of float64 = 41EFFFFFFFE00000
2) SLAW_NOT_NUMERIC
3) SLAW_NOT_NUMERIC
